### PR TITLE
fix(aria-hidden-focus): mark as needs review if a modal is open

### PR DESCRIPTION
**Pull request by straker**
The rule will now return as needs review if there are focusable elements present while what we believe to be a modal is open. On the react example, it will now return as needs review rather than failing when the modal is open that traps focus https://react-bootstrap.github.io/components/modal/#modals-live.

Closes issue: #1482

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**

- [x] Follows the commit message policy, appropriate for next version
- [x] Code is reviewed for security

### CODE REVIEW

File: lib/commons/dom/is-modal-open.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	let fullWidthElement;
+	stacks.forEach(stack => {
+		// the modal element may not be the top-most element of the stack due to
+		// how it's written, so we'll look for it in the first half
+		for (let i = 0; i < stack.length / 2; i++) {
+			const style = window.getComputedStyle(stack[i]);
+			const modal =
+				parseInt(style.width, 10) >= percentWidth &&
+				parseInt(style.height, 10) >= percentHeight &&
+				(style.position === 'absolute' || style.position === 'fixed');
```
Comment (WilcoFiers) [inline lib/commons/dom/is-modal-open.js]: Should we look at pointer-events here too? If this has pointer-events:none on it it's likely not an overlay.

File: lib/commons/dom/is-modal-open.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	// four corners and one from the middle of the viewport (total of 5). if each
+	// stack contains an element whose width/height is >= 75% of the screen as
+	// the top most element, we found a "could be a modal"
+	const viewport = dom.getViewportSize(window);
+	const percentWidth = viewport.width * modalPercent;
+	const percentHeight = viewport.height * modalPercent;
+	const x = (viewport.width - percentWidth) / 2;
+	const y = (viewport.height - percentHeight) / 2;
+
+	const points = [
```
Comment (WilcoFiers) [inline lib/commons/dom/is-modal-open.js]: Wondering if we shouldn't also look at the center point.

File: lib/commons/dom/is-modal-open.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+	const stacks = points.map(point => {
+		return document.elementsFromPoint(point.x, point.y);
+	});
+
+	let fullWidthElement;
+	stacks.forEach(stack => {
+		// the modal element may not be the top-most element of the stack due to
+		// how it's written, so we'll look for it in the first half
+		for (let i = 0; i < stack.length / 2; i++) {
```
Comment (WilcoFiers) [inline lib/commons/dom/is-modal-open.js]: This "half the stack" thing feels pretty arbitrary. Why did you choose this?

File: test/commons/dom/is-modal-open.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	);
+
+	it('returns true if there is a visible absolutely positioned element with >= 75% width/height', function() {
+		fixtureSetup(
+			'<div style="position: absolute; top: 0; bottom: 0; left: 0; right: 0">Modal</div>'
+		);
+		assert.isTrue(isModalOpen());
+	});
+
+	it('returns true if there is a visible absolutely positioned element with >= 75% width/height', function() {
```
Comment (WilcoFiers) [inline test/commons/dom/is-modal-open.js]: This test is a duplicate of the one above it.

File: test/commons/dom/is-modal-open.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		fixtureSetup('<div>Modal</div>');
+		assert.isUndefined(isModalOpen());
+	});
+
+	it('returns undefined if there is a hidden element with role=dialog', function() {
+		fixtureSetup('<div role="dialog" style="display: none">Modal</div>');
+		assert.isUndefined(isModalOpen());
+	});
+
+	it('returns undefined if there is a hidden element with aria-modal=undefined', function() {
```
Comment (WilcoFiers) [inline test/commons/dom/is-modal-open.js]:
```suggestion
	it('returns undefined if there is a hidden element with aria-modal=true', function() {
```

File: lib/commons/dom/is-modal-open.js [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	if (
+		fullWidthElement &&
+		stacks.every(stack => Array.from(stack).includes(fullWidthElement))
+	) {
+		axe._cache.set('isModalOpen', true);
+		return true;
+	}
+
+	axe._cache.set('isModalOpen', undefined);
+	return undefined;
```
Comment (WilcoFiers) [inline lib/commons/dom/is-modal-open.js]: Why is this returning true | undefined instead of just true | false?

Comment (straker) [reply lib/commons/dom/is-modal-open.js]: I thought about returning false but I didn't want it to seem like the test was accurate and could say "false" meant a modal was not open. Undefined kinda help make the check feel more uncertain that we're not even sure if a modal is open or not.

**Pull request merged**